fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/ContainmentIndexImpl.java

```diff
@@ -0,0 +1,189 @@
+/*
+ * Licensed to DuraSpace under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional information
+ * regarding copyright ownership.
+ *
+ * DuraSpace licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except in
+ * compliance with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.fcrepo.kernel.impl;
+
+import org.fcrepo.kernel.api.ContainmentIndex;
+import org.fcrepo.kernel.api.Transaction;
+import org.fcrepo.kernel.api.models.FedoraResource;
+import org.springframework.jdbc.datasource.DriverManagerDataSource;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+import java.sql.Connection;
+import java.sql.DatabaseMetaData;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.stream.Stream;
+
+/**
+ * @author peichman
+ */
+public class ContainmentIndexImpl implements ContainmentIndex {
+    @Inject
+    private DriverManagerDataSource containmentIndexDataSource;
+
+    private Connection conn;
+
+    private PreparedStatement childrenQuery;
+
+    private PreparedStatement insertChild;
+
+    private PreparedStatement deleteChild;
+
+    private static final String TABLE_NAME = "resources";
+
+    private static final String FEDORA_ID_COLUMN = "fedoraId";
+
+    private static final String PARENT_COLUMN = "parent";
+
+    private static final String RESOURCES_TABLE_DDL = "CREATE TABLE " + TABLE_NAME + " (" +
+            FEDORA_ID_COLUMN + " text PRIMARY KEY, " +
+            PARENT_COLUMN + " text);";
```

Comment:
Thinking about this some more. Should the table have a row for transaction id? Then you can just add rows with a transaction id and then either delete them for a rollback or update them (to remove the transaction id) for a commit. Then the default check is all rows without a transaction id or where the transaction id matches the provided one.

Reply:
Not sure how we deal with "deleted" rows in a transaction....

Reply:
@whikloj I like that idea. To track deleted relationships, perhaps we add an additional column that flags a row as "to be deleted" in the scope of the transaction.